### ISSUE TITLE
I need some insight about the training process

### ISSUE THREAD

**Issue by FranciscoGomez90** [affinelayer/pix2pix-tensorflow#159]
Hi there,
I am relatively new to the GAN world, but quite experienced with deep learning in general.
Although I have read a lot about GANs, and specifically about CycleGans, there are some details that I cannot properly grasp. I hope some of you could please provide some insight about the following question.
About the training process, AFAIK the "discriminator" is fed with real and fake images. Is this done separately? I mean, first we feed the network with real data and compute the discriminator real loss, and then we do the same with fake images in order to compute the discriminator fake loss. Am I right so far? Is the same instance of the discriminator used for real and fake data (same weights)? Are the weights updated separately for real and then for fake?

Could you please explain how the losses are computed and what are the actual input and output of the discriminator?

Thank you.

**Comment by HenryCWoo** [comment]
It looks like the input to the discriminator is fed in pairs: (input + target) and (input + generated image). By pairs, I mean quite literally the input image is glued to the target/generated image side-by-side to produce a single image.

To answer your question about how the loss is computed, we need to get more specific. The output uses a sigmoid activation for a float between 0 (fake) and 1 (real). So now you have 2 floats. 
Take the sum of their logs:
`log(real_output) + log(1-fake_output)`
 and then negate it so that the final sum is positive.
That is for one instance though. Now think in terms of batches, this is computed for all instances then the average is taken to get the final loss from batches. (See line 448 in pix2pix.py.)
So yes, losses are computed separately, but the weight/instance of the discriminator used for real and fake is the same.

I'm also relatively new to the GAN world, so please let me know if I misunderstood something!
